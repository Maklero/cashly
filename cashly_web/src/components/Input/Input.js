import PropTypes from "prop-types";
import styled, { css } from "styled-components";


const StyledWrapper = styled.div`
    width: 256px;
    display: flex;
    flex-direction: column;

    ${({ fullWidth }) => fullWidth && css`width: 100%;`}

    &:focus-within {
        ${".inputLabel"} {
            color: white;
            background-color: ${({ theme }) => theme.colors.primary400};
        }
    }
`;

const StyledInput = styled.input`
    width: 100%;
    outline: none;
    line-height: 1.5;
    padding: 5px 15px;
    background-color: white;
    font-size: ${({ theme }) => theme.fontSizes.h5};
    font-family: ${({ theme }) => theme.fontFamily};
    color: ${({ theme }) => theme.colors.primary400};
    font-weight: ${({ theme }) => theme.fontWeights.regular};
    border: 1px solid ${({ theme }) => theme.colors.primary400};
    border-bottom-left-radius: ${({ isError }) => isError ? "0" : "2px"};
    border-bottom-right-radius: ${({ isError }) => isError ? "0" : "2px"};
    border-top-left-radius: ${({ labelText }) => labelText ? "0" : "2px"};
    border-top-right-radius: ${({ labelText }) => labelText ? "0" : "2px"};
`;

const StyledLabel = styled.label`
    padding: 3px 8px;
    width: fit-content;
    background-color: white;
    border-radius: 2px 2px 0 0;
    border-bottom: none !important;
    font-size: ${({ theme }) => theme.fontSizes.h6};
    color: ${({ theme }) => theme.colors.primary400};
    transition: ${({ theme }) => theme.defaultTransition};
    font-weight: ${({ theme }) => theme.fontWeights.semiBold};
    border: 1px solid ${({ theme }) => theme.colors.primary400};
`;

const StyledError = styled.p`
    color: white;
    padding: 6px 8px;
    text-align: left;
    word-break: break-word;
    border-radius: 0 0 2px 2px;
    font-size: ${({ theme }) => theme.fontSizes.h6};
    background-color: ${({ theme }) => theme.colors.red400};
    font-weight: ${({ theme }) => theme.fontWeights.semiBold};
`;

function Input(props) {
    const { className, fullWidth, labelText, error } = props;
    const isError = Boolean(error);

    return (
        <StyledWrapper fullWidth={fullWidth}>
            {labelText && (
                <StyledLabel className="inputLabel">
                    {labelText}
                </StyledLabel>
            )}
            <StyledInput
                className={className}
                isError={isError}
                {...props} 
            />
            {error && <StyledError>{error}</StyledError>}
        </StyledWrapper>
    );
} 

Input.propTypes = {
    error: PropTypes.string,
    fullWidth: PropTypes.bool,
    labelText: PropTypes.string
};

Input.defaultProps = { fullWidth: false };

export default Input;