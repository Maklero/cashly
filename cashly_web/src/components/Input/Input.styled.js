import styled, { css } from 'styled-components'

export const StyledContainer = styled.div`
    width: 256px;
    row-gap: 8px;
    display: flex;
    flex-direction: column;

    input[type='date'],
    input[type='color'],
    select {
        height: 53.6px;
    }

    input[type='color'] {
        padding: 0;
    }

    &:focus-within {
        .labelText {
            color: ${({ theme, isError }) =>
                isError ? theme.colors.red400 : theme.colors.primary400};
        }
    }

    ${({ theme, isError }) =>
        isError &&
        css`
            .labelText {
                color: ${theme.colors.red400};
            }

            .labelSeparator {
                background-color: ${theme.colors.red400};
            }

            .input {
                border-color: ${theme.colors.red400};
            }

            .input:focus {
                border-color: ${theme.colors.red400};
            }
        `}

    ${({ fullWidth }) =>
        fullWidth &&
        css`
            width: 100%;
        `}
`

export const StyledInput = styled.input`
    width: 100%;
    outline: none;
    padding: 16px;
    font-weight: 400;
    font-size: 0.938rem;
    background-color: #fbfbfb;
    border: 2px solid #cdcdcd;
    font-family: ${({ theme }) => theme.fontFamily};
    color: ${({ theme }) => theme.colors.primary400};

    &:focus {
        border-color: ${({ theme }) => theme.colors.primary400};
    }

    &:placeholder {
        color: #eaeaea;
    }
`

export const StyledLabelContainer = styled.div`
    display: flex;
    column-gap: 4px;
    align-items: center;
    flex-direction: row;
`

export const StyledLabelSeparator = styled.span`
    width: 8px;
    height: 1px;
    background-color: #858585;
`

export const StyledLabelText = styled.span`
    color: #858585;
    font-weight: 700;
    width: fit-content;
    font-size: 0.75rem;
`
