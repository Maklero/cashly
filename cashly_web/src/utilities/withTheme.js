import { ThemeProvider } from "styled-components";
import { theme } from "../styles";

const withTheme = (Component) => {
    return (
        <ThemeProvider theme={theme}>
            <Component />
        </ThemeProvider>
    );
}

export default withTheme;