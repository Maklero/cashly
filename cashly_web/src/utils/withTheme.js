import { ThemeProvider } from 'styled-components'

import { theme } from '@/styles'

export const withTheme = (component) => (
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
)
