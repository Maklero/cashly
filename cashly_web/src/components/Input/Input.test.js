import { render, screen } from "@testing-library/react";
import withTheme from "../../utilities/withTheme";
import Input from "./Input";

test("renders input component", () => {
    const PreparedInputComponent = () => {
        return <Input type="text" />;
    }

    const { container } = render(withTheme(PreparedInputComponent));

    const inputElm = screen.getByTestId("input");

    expect(container.firstChild).toMatchSnapshot();
    expect(inputElm).toHaveAttribute("type", "text");
})

test("renders input component with label", () => {
    const PreparedInputComponent = () => {
        return <Input type="text" labelText="Example label" />;
    }

    const { container } = render(withTheme(PreparedInputComponent));

    const inputElm = screen.getByTestId("input");
    const labelElm = screen.getByTestId("label");

    expect(inputElm).toHaveAttribute("type", "text");
    expect(labelElm).toHaveTextContent("Example label");
    expect(container.firstChild).toMatchSnapshot();
})