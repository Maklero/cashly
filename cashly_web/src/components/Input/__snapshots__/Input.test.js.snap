// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders input component 1`] = `
<div
  class="sc-bdvvtL gnllsT"
>
  <input
    class="sc-dkPtRN kHWYqx"
    data-testid="input"
    type="text"
  />
</div>
`;

exports[`renders input component with label 1`] = `
<div
  class="sc-bdvvtL gnllsT"
>
  <label
    class="sc-gsDKAQ iOutxc"
    data-testid="label"
  >
    Example label
  </label>
  <input
    class="sc-dkPtRN kHWYqx"
    data-testid="input"
    type="text"
  />
</div>
`;
